text/0047-random-assignment-of-availability-chunks.md

```diff
@@ -0,0 +1,283 @@
+have as little overlap as possible for different paras scheduled on that relay parent.
+
+In other words, we want a uniformly distributed, deterministic mapping from `ValidatorIndex` to `ChunkIndex` per block
+per scheduled para.
+
+#### Proposed runtime API
+
+The mapping function should be implemented as a runtime API, because:
+
+1. it enables further atomic changes to the shuffling algorithm.
+1. it enables alternative client implementations (in other languages) to use it
+1. considering how critical it is for parachain consensus that all validators have a common view of the Validator->Chunk
+mapping, this mitigates the problem of third-party libraries changing the implementations of the `ChaCha8Rng` or the `rand::shuffle`
+that could be introduced in further versions. This would stall parachains if only a portion of validators upgraded the node.
+
+
+Pseudocode:
+
+```rust
+pub fn get_chunk_index(
+  n_validators: u32,
+  validator_index: ValidatorIndex,
+  relay_parent: Hash,
+  para_id: ParaId
+) -> ChunkIndex {
+  let threshold = systematic_threshold(n_validators); // Roughly n_validators/3
+  let seed = derive_seed(relay_parent);
+  let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);
+  let mut chunk_indices: Vec<ChunkIndex> = (0..n_validators).map(Into::into).collect();
+  chunk_indices.shuffle(&mut rng);
+
+  let seed = derive_seed(hash(para_id));
+  let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);
+  let para_start_pos = rng.gen_range(0..n_validators);
+
+  chunk_indices[(para_start_pos + validator_index) % n_validators]
+}
+```
+
+Additionally, so that client code is able to efficiently get the mapping from the runtime, another API will be added
+for retrieving chunk indices in bulk for all validators at a given block and core:
+
+```rust
+pub fn get_chunk_indices(
+  n_validators: u32,
+  relay_parent: Hash,
+  para_id: ParaId
+) -> Vec<ChunkIndex> {
+  let threshold = systematic_threshold(n_validators); // Roughly n_validators/3
+  let seed = derive_seed(relay_parent);
+  let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);
+  let mut chunk_indices: Vec<ChunkIndex> = (0..n_validators).map(Into::into).collect();
+  chunk_indices.shuffle(&mut rng);
+
+  let seed = derive_seed(hash(para_id));
+  let mut rng: ChaCha8Rng = SeedableRng::from_seed(seed);
+
+  let para_start_pos = rng.gen_range(0..n_validators);
+
+  chunk_indices
+    .into_iter()
+    .cycle()
+    .skip(para_start_pos)
+    .take(n_validators)
+    .collect()
+}
+```
+
+#### Upgrade path
+
+Considering that the Validator->Chunk mapping is critical to para consensus, the change needs to be enacted atomically
+via governance, only after all validators have upgraded the node to a version that is aware of this mapping.
+It needs to be explicitly stated that after the runtime upgrade and governance enactment, validators that run older
+client versions that don't support this mapping will not be able to participate in parachain consensus.
```

Comment:
Does this need to be the case? As discussed, I think they should be able to still do the recovery from non systemic chunks.

Reply:
yeah but they need to know what chunks map to where.

Reply:
yep in distribution, for fetching I think the mapping ValidatorIndex -> ChunkIndex can be unknown, without any real issues.